### DISCUSSION TITLE
Custom Interfaces, migration from pre-record builder libs, etc.

### DISCUSSION THREAD

**Discussion by Randgalt** [Randgalt/record-builder#21] [Ideas]
Discussion on extensions to the Record Builder library to help with migration from pre-record libraries such as Immutables, AutoValue, lombok, etc.

**Comment by jedvardsson** [comment]
Thanks for a great tool! Any plans or ideas of how to have the generated builder to implement custom interfaces?

```
interface Entity<E extends Entity<E, B>, B extends Entity.Builder<E, B>> {
    String id();

    interface Builder<E extends Entity<E, B>, B extends Builder<E, B>> {
        B id(String name);
        E build();
    }
}

@RecordInterface
interface User extends Entity<User, User.Builder> {
    @RecordBuilderInterface
    interface Builder extends Entity.Builder<User, Builder> {
    }
}
```

**Comment by Randgalt** [comment]
This is similar to what is asked for here: https://github.com/Randgalt/record-builder/issues/20

I'm reticent to add too many additional features. I'm trying to stay close to what Oracle has said will be coming in the future. Can you give more detail on what would get generated in this case?

**Reply by jedvardsson** [reply]
I agree with you to keep things as simple as possible. 

Basically it would be a way to add additional methods to the generated builder. In this way RecordBuilder would be almost as flexible as AutoValue. 

In AutoValue abstract methods of getter or setter style are automatically implemented. Non-abstract methods are left untouched. In this way AutoValue fills in the blanks so to say.

For the example above RecordBuilder would generate something like:
```
record UserRecord(String id) implements User {
}
public class UserRecordBuilder implements User.Builder {
   private String id;
   @Generated("io.soabase.recordbuilder.core.RecordBuilder")
   String id() {
       return id;
   }
   @Generated("io.soabase.recordbuilder.core.RecordBuilder")
   UserRecordBuilder id(String id) {
        this.id = id;
        return this;
   }
}
```

I'm not sure if I make sense?

**Comment by TomerFi** [comment]
Great tool!
Thank you very much.
This is going to make the transition from auto-value to record classes much less painful. 

> I'm reticent to add too many additional features. I'm trying to stay close to what Oracle has said will be coming in the future.

Where might one read about record classes future, any link you can share please?

**Reply by Randgalt** [reply]
This is a good one (though there's a huge number of posts, mail messages, links, etc.): https://github.com/openjdk/amber-docs/blob/master/eg-drafts/reconstruction-records-and-classes.md

**Comment by Randgalt** [comment]
On this subject - I'd be happy to extend the library to make transition from AutoValue, Immutables and Lombok (I'll need it myself). But, it needs to be done carefully. I really don't want this library to turn into bloatware.

**Reply by jedvardsson** [reply]
Great. I’ll be happy to help if I can. I’ll spend some thought on how it could be done.

**Reply by Randgalt** [reply]
We need a proposal first - examples from some of these other libraries and what it might look like in Record Builder (without worrying about writing an implementation)
